rust/threshold-schnorr/src/schnorr_example_rust/src/lib.rs

```diff
@@ -0,0 +1,223 @@
+    pub signature: Vec<u8>,
+}
+
+thread_local! {
+    static STATE: RefCell<String> = RefCell::new("aaaaa-aa".to_string());
+}
+
+#[update]
+async fn for_test_only_change_management_canister_id(id: String) -> Result<(), String> {
+    let _ = CanisterId::from_text(&id).map_err(|e| panic!("invalid canister id: {}: {}", id, e));
+    STATE.with_borrow_mut(move |current_id| {
+        println!(
+            "Changing management canister id from {} to {id}",
+            *current_id
+        );
+        *current_id = id;
+    });
+    Ok(())
+}
+
+#[update]
+async fn public_key(algorithm: SchnorrAlgorithm) -> Result<PublicKeyReply, String> {
+    let request = ManagementCanisterSchnorrPublicKeyRequest {
+        canister_id: None,
+        derivation_path: vec![ic_cdk::api::caller().as_slice().to_vec()],
+        key_id: SchnorrKeyIds::TestKeyLocalDevelopment.to_key_id(algorithm),
+    };
+
+    let (res,): (ManagementCanisterSchnorrPublicKeyReply,) =
+        ic_cdk::call(mgmt_canister_id(), "schnorr_public_key", (request,))
+            .await
+            .map_err(|e| format!("schnorr_public_key failed {}", e.1))?;
+
+    Ok(PublicKeyReply {
+        public_key_hex: hex::encode(&res.public_key),
+    })
+}
+
+#[update]
+async fn sign(message: String, algorithm: SchnorrAlgorithm) -> Result<SignatureReply, String> {
+    let internal_request = ManagementCanisterSignatureRequest {
+        message: message.as_bytes().to_vec(),
+        derivation_path: vec![ic_cdk::api::caller().as_slice().to_vec()],
+        key_id: SchnorrKeyIds::TestKeyLocalDevelopment.to_key_id(algorithm),
+    };
+
+    let (internal_reply,): (ManagementCanisterSignatureReply,) =
+        ic_cdk::api::call::call_with_payment(
+            mgmt_canister_id(),
+            "sign_with_schnorr",
+            (internal_request,),
+            25_000_000_000,
+        )
+        .await
+        .map_err(|e| format!("sign_with_schnorr failed {e:?}"))?;
+
+    Ok(SignatureReply {
+        signature_hex: hex::encode(&internal_reply.signature),
+    })
+}
+
+#[query]
+async fn verify(
+    signature_hex: String,
+    message: String,
+    public_key_hex: String,
+    algorithm: SchnorrAlgorithm,
+) -> Result<SignatureVerificationReply, String> {
+    let sig_bytes = hex::decode(&signature_hex).expect("failed to hex-decode signature");
+    let msg_bytes = message.as_bytes();
+    let pk_bytes = hex::decode(&public_key_hex).expect("failed to hex-decode public key");
+
+    match algorithm {
+        SchnorrAlgorithm::Bip340Secp256k1 => {
+            verify_bip340_secp256k1(&sig_bytes, msg_bytes, &pk_bytes)
+        }
+        SchnorrAlgorithm::Ed25519 => verify_ed25519(&sig_bytes, &msg_bytes, &pk_bytes),
+    }
+}
+
+fn verify_bip340_secp256k1(
+    sig_bytes: &[u8],
+    msg_bytes: &[u8],
+    secp1_pk_bytes: &[u8],
+) -> Result<SignatureVerificationReply, String> {
+    assert_eq!(secp1_pk_bytes.len(), 33);
+    assert_eq!(sig_bytes.len(), 64);
```

Comment:
`assert_eq!(sig_bytes.len(), 64);` was actually quite redundant. I left `assert_eq!(secp1_pk_bytes.len(), 33);` there because it might be useful for debugging s.t. it's more obvious that we expect non-truncated key there. It's easier in that way to copy-paste the key in the UI and not care about what to truncate.